### ISSUE TITLE
Else block always runs

### ISSUE THREAD

**Issue by kylebarron** [kylebarron/stata_kernel#150]
@mcaceresb 
This is because we run other commands after the `if` clause gets sent. Not really sure the best way to fix this. In Jupyter Notebook/Atom you can just be careful to only send the entire if/else block. In Jupyter console, you can fix this by wrapping the entire block in a pair of braces.

![image](https://user-images.githubusercontent.com/15164633/45238813-1836cc00-b2b1-11e8-8e3a-9b4dd06eee1c.png)
![image](https://user-images.githubusercontent.com/15164633/45238857-3c92a880-b2b1-11e8-9879-8122d7acef0a.png)

Console workaround:
![image](https://user-images.githubusercontent.com/15164633/45239008-bd51a480-b2b1-11e8-9136-f11335e8c47e.png)

**Comment by kylebarron** [comment]
This is helpful for the console, but the behavior is still suboptimal in Hydrogen, since it doesn't use `is_complete` messages at all..
![peek 2018-09-07 15-27](https://user-images.githubusercontent.com/15164633/45239253-9d6eb080-b2b2-11e8-845e-a84e546442e4.gif)

**Comment by kylebarron** [comment]
The only way to get around this in hydrogen right now is to manually select the if/else block. For now I should at least put this in the limitations page in the docs.

![peek 2018-09-07 15-28](https://user-images.githubusercontent.com/15164633/45239279-b24b4400-b2b2-11e8-962e-779008cd76da.gif)

**Comment by mcaceresb** [comment]
Wait, how does it do it for python, then?

**Comment by kylebarron** [comment]
It doesn't. Since Hydrogen is 100% language-agnostic, it uses only general syntax elements like indentation and the `foldEndMarker` designated by the language's syntax highlighting package.

Recently a separate [python-specific add-on package](https://github.com/nikitakit/hydrogen-python) was developed to expand code lines. See https://github.com/nikitakit/hydrogen-python/pull/10 for the difficulty of correctly implementing this feature.

This is something Hydrogen struggles with in general because of its language-independentness... Here's the same issue in Javascript.
![peek 2018-09-07 15-47](https://user-images.githubusercontent.com/15164633/45240093-66e66500-b2b5-11e8-91dd-f822c5745190.gif)

I think it's something that could be improved in Hydrogen... If you had a suggestion I might mention it to the other maintainers and try to create a PR for it.

**Comment by mcaceresb** [comment]
> and the foldEndMarker designated by the language's syntax highlighting package.

I thought you were in charge of that, or is it too much of a pain to implement? You can nest if/else in an ifelse fold if if is followed by else in the next line?

**Comment by kylebarron** [comment]
I am. And I [edited it recently](https://github.com/kylebarron/language-stata/commit/9a41b92e45a87d2843030488e08c5663d4313a03) to include the ending `}` in Hydrogen.

But it's just a plain regex setting (that a core Atom package then implements) and Atom can't do regexes across lines anyways. So I don't think it's possible to say "`}` ends folds except when `else {` or `else if {` appears on the following line"

**Comment by kylebarron** [comment]
In Javascript, you could put `}(?!\s* else)` as the regex and plausibly get the whole if/else clause. But not sure how to do it when they're always on separate lines.

**Comment by mcaceresb** [comment]
Multi-line regex?

**Comment by kylebarron** [comment]
> Atom can't do regexes across lines anyways

At least not for syntax highlighting. I doubt it allows it here, though I could try it out.

**Comment by mcaceresb** [comment]
Wait, this is odd. Python has multi-line strings. Doesn't it highlight them?

Conceptually, `if` should start an if-else environment ended by a blank line or an else block, right? I reckon that's not how it's written?

**Comment by kylebarron** [comment]
> Wait, this is odd. Python has multi-line strings. Doesn't it highlight them?

Yes, but the token that it looks for in the regular expression never spans a newline. I.e. the `"""` will be on one line and another, and Atom can create regions between the lines, but the token can never span lines.

That's why it's impossible, for example, to highlight Setext headers in Markdown
https://github.com/burodepeper/language-markdown/issues/192

**Comment by mcaceresb** [comment]
Actually, it is ended by an else block or a line containing anything that doesn't start with `else`

**Comment by mcaceresb** [comment]
But that is not the same, tho, right? Because you can look for environment delimiters...at least that is what I'm assuming.

**Comment by mcaceresb** [comment]
The markdown example has no delimiter to start the header, just to end it.

**Comment by kylebarron** [comment]
With the syntax highlighting yes. But not the fold end marker to my knowledge. I haven't tried but I'd be surprised.

**Comment by kylebarron** [comment]
What's a simple regex to try?
```
}\s*\n\s*(?!else)
```

**Comment by mcaceresb** [comment]
Something like that, yeah. Remember that you can have `if` w/o brackets and else with brackets after afaik

**Comment by kylebarron** [comment]
It actually works when not using brackets:
![image](https://user-images.githubusercontent.com/15164633/45243113-9ea6da00-b2c0-11e8-89f9-39d0c16fc4df.png)

I suppose the main reason is that when there are brackets I run the code with `include`

**Comment by kylebarron** [comment]
So `include` must break Stata's record of `if/else` somehow

**Comment by kylebarron** [comment]
The only case in which it _doesn't_ work is when `if` uses brackets.

Why did we end up using include anyways?

![peek 2018-09-07 17-10](https://user-images.githubusercontent.com/15164633/45243217-f2b1be80-b2c0-11e8-9b18-a4b117f0ae93.gif)

**Comment by kylebarron** [comment]
Oh right we use `include` so that commands will stop running after one creates an error. Which is good...
